### ISSUE TITLE
BUG - broken UTF 8 encoded content output

### ISSUE THREAD

**Issue by OSPanel** [curl/curl#9841]

**Comment by OSPanel** [comment]
I can't compile the binary, but I can check the test binary if needed.

**Comment by vszakats** [comment]
@OSPanel: You can find fresh Unicode builds here: https://ci.appveyor.com/project/curlorg/curl-for-win/builds/45248995/artifacts

**Comment by OSPanel** [comment]
Thanks. I downloaded and tested these x86+x64 builds (dated 10/26/2022), and this bug was found there :-(

**Comment by vszakats** [comment]
I freshly made these builds today, right before posting here. They have Unicode enabled (check `curl -V`).

[ For reproducibility the dates are always the same for a given curl version number. ]

**Comment by OSPanel** [comment]
I used this link: https://ci.appveyor.com/project/curlorg/curl-for-win/builds/45248995/artifacts
Release-Date: 2022-10-26
```
url -V
curl 7.86.0 (x86_64-w64-mingw32) libcurl/7.86.0 OpenSSL/3.0.5 (Schannel) zlib/1.2.13 brotli/1.0.9 zstd/1.5.2 WinIDN libssh2/1.10.0 nghttp2/1.50.0 ngtcp2/0.10.0 nghttp3/0.7.1 libgsasl/2.2.0
Release-Date: 2022-10-26
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp ws wss
Features: alt-svc AsynchDNS brotli gsasl HSTS HTTP2 HTTP3 HTTPS-proxy IDN IPv6 Kerberos Largefile libz MultiSSL NTLM SPNEGO SSL SSPI threadsafe TLS-SRP Unicode UnixSockets zstd
```

**Comment by vszakats** [comment]
@OSPanel: That's the correct one indeed, and thanks for testing it!

Based on your tests, this issue is present in both Unicode and non-Unicode builds.

**Comment by bagder** [comment]
I don't see an explanation exactly what the symptoms of this error is?

Is it perhaps code from this commit that causes it? 5bfaa86ceb3c2a9ac474a928e748c4a86a703b33

**Comment by vszakats** [comment]
I'm making an [attempt](https://github.com/curl/curl-for-win/commit/b39b6b457f6462e21c4db764a109dc55a4ede56c) to revert this patch in the test build here: https://ci.appveyor.com/project/curlorg/curl-for-win/builds/45254038

Even though the notion of the patch seems correct to me and changing terminal codepage on the fly was fragile/broken at least before Windows 10.

**Comment by OSPanel** [comment]
> I'm making an [attempt](https://github.com/curl/curl-for-win/commit/b39b6b457f6462e21c4db764a109dc55a4ede56c) to revert this patch in the test build here: https://ci.appveyor.com/project/curlorg/curl-for-win/builds/45254038
> 
> Even though the notion of the patch seems correct to me and changing terminal codepage on the fly was fragile/broken at least before Windows 10.

Thanks. This test build works well, the bug we are discussing has not been detected.

**Comment by vszakats** [comment]
It means that 5bfaa86ceb3c2a9ac474a928e748c4a86a703b33 is causing the difference.

According to past experience fixing this exact issue in a different project, and this earlier Python [Issue](https://github.com/python/cpython/issues/45943), the `WriteConsoleW()` call _should_ be correct here. Maybe the CP/encoding translation before it has issues? E.g. when then input stream has a multi-byte character which is processed in two passes, breaking the char into two halves?

**Comment by jay** [comment]
Yes, I think the OP called it correctly. We are passing an incomplete UTF-8 encoding to MultiByteToWideChar more than one time (up to 4 times since there are up to 4 bytes in UTF-8 encoded characters). That results in an invalid character � (U+FFFD) output for each incomplete UTF-8 encoded character.

I cannot reproduce with the OP's URL but I have no doubt this issue is valid. Let's take the first mangled output for example:
~~~
Команда доступна только для пассивных (не подлежащих запуску) модулей!
                                    ^
Команда доступна только для пассивны�� (не подлежащих запуску) модулей!
~~~
A caret is pointing to the `х` that becomes mangled. Though it looks like the regular ascii x (U+0078) in the screenshot it's actually a cyrillic х (U+0445) which is represented in UTF-8 encoding as `D1 85`. It looks likely that MultiByteToWideChar received D1 at the end of one buffer and then 85 at the start of the next buffer.

I think that can be fixed by delaying the conversion and output for incomplete Unicode UTF-8 encoded characters until they are complete.  I had at one point written a UTF-8 correctness function for libcurl for other reasons but I don't think we added it. I will see if maybe some of that logic applies here.

Also, the comments in those python threads about WriteConsole long output failing cconcern me. The MS documentation currently says _"If the total size of the specified number of characters exceeds the available heap, the function fails with ERROR_NOT_ENOUGH_MEMORY"_ but according to the python threads it used to reference a specific shared 64k heap that may have an arbitrary size available. That is something we'd need to actually take into consideration though I wonder why they changed the documentation. I will look into this as well.

**Comment by bagder** [comment]
Any progress on this @jay? It feels like a rather complicated dilemma.

**Comment by jay** [comment]
I think it's possible. I was focused on other bugs but I will take a look at it tomorrow.

**Comment by OSPanel** [comment]
https://github.com/curl/curl/commit/5bfaa86ceb3c2a9ac474a928e748c4a86a703b33 - I assume that this is a bad fix. You cannot output UTF 8 to the console if the current encoding is different from 65001. Therefore, the old code worked well, it cached the current encoding in the console and returned it after the output was completed. This is a normal approach. And if someone has low-quality fonts installed that do not support UTF 8, and because of this their console looks bad, so it's not a CURL problem, it's a problem of the user and his chosen font. Just return the old code by canceling this commit.

**Comment by jay** [comment]
I've been busy and haven't had a chance to work on it. curl outputs UTF-16 to the console, it is the conversion UTF-8 => UTF-16 which is failing in this case. The user's font may not be able to display some valid Unicode characters that are being output but IMO that is not a curl issue.

**Comment by OSPanel** [comment]
Yes, before the commit https://github.com/curl/curl/commit/5bfaa86ceb3c2a9ac474a928e748c4a86a703b33 was applied, curl's work in the console depended only on the user's font, which is why I suggest returning the old code. Now, after that commit was accepted, the output to the console is completely broken. Maybe you will return the old code until you get around to changing something?

**Comment by vszakats** [comment]
The old code was broken in a different way, e.g. it could leave the console in an unstable state (leading to a console crash).

**Comment by pavlusha23** [comment]
Any news on fixes yet?

**Comment by OSPanel** [comment]
`curl https://wttr.in/London?lang=nl`
As it turned out, CURL distorts not only the text output, but even the output of simple graphic lines. Reproducibility of the problem - every first or second request (~70% of requests). 
<a href="https://ibb.co/SJ03vDw"><img src="https://i.ibb.co/Km7rycx/er2.png" alt="er2" border="0"></a>

**Comment by dfandrich** [comment]
That last problem doesn't seem related to me. Also, how do you know it's a curl problem and not a problem with the server, network, network stack or terminal?

**Comment by OSPanel** [comment]
> That last problem doesn't seem related to me. Also, how do you know it's a curl problem and not a problem with the server, network, network stack or terminal?

The problem is related to encoding, so it is not always reproduced, it depends on the output data. When using the city of London, the problem is no longer visible, but from the 10th time I was able to pick up a location where the reproducibility of the breach is 100% right now (the weather will change during your viewing and you may have to pick up another location):

`curl.exe https://wttr.in/Malta?lang=nl`
`curl.exe https://wttr.in/pitsburg?lang=nl`

<a href="https://ibb.co/dKVQ8GF"><img src="https://i.ibb.co/q1cRTkK/er1.png" alt="er1" border="0"></a>
<a href="https://ibb.co/wJzfRzj"><img src="https://i.ibb.co/q0rvdrV/er2.png" alt="er2" border="0"></a>

**Comment by dfandrich** [comment]
This is very strange. I can reproduce it in a Linux terminal in wine! I'm using
the official curl 32-bit build:

curl 7.87.0 (i686-w64-mingw32) libcurl/7.87.0 OpenSSL/3.0.8 (Schannel) zlib/1.0.
Release-Date: 2022-12-21
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps sq
Features: alt-svc AsynchDNS brotli gsasl HSTS HTTP2 HTTP3 HTTPS-proxy IDN IPv6dK

But what's strangest is that when I redirect the output to a file and cat it
outside wine, it's always fine. But when I run curl in the Linux terminal with
wine, that line is corrupted probably 80% of the time ATM. Could it be an issue
with output when curl detects a terminal and not when it's directed to a file?

**Comment by OSPanel** [comment]
> Could it be an issue with output when curl detects a terminal and not when it's directed to a file?

So it is, only the output to the console is broken. Read the posts above here.

**Comment by dfandrich** [comment]
I have another clue. By looking at --trace data at the location of the
corruption, you can see this:

```
1690: 94 80 e2 94 80 e2 94 80 e2 94 80 e2 94 80 e2 94 ................
16a0: 80 e2 94 80 e2 94 80 e2 94 b4 e2 94 80 e2 94 80 ................
16b0: e2 94 80 e2 94 80 e2 94 80 e2 94 80 e2 94 80 e2 ................
16c0: 94 80 e2 94 80 e2 94 80 e2 94 80 e2 94 80 e2 94 ................
16d0: 80 e2 94 80 e2 94 80 e2 94 80 e2 94 80 e2 94 80 ................
16e0: e2 94 80 e2 94 80 e2 94 80 e2 94 80 e2 94 80 e2 ................
16f0: 94 80 e2 94 80 e2 94 80 e2 94 80 e2 94 80 e2 94 ................
1700: 80 e2 94 80 e2 94 b4 e2 94 80 e2 94 80 e2 94 80 ................
1710: e2 94 80 e2 94 80 e2 94 80 e2 94 80 e2 94 80 e2 ................
1720: 94                                              .
== Info: [CONN-0-0][CF-SSL] TLSv1.2 (IN), TLS header, Supplemental data (23):
<= Recv SSL data, 5 bytes (0x5)
0000: 17 03 03 0b c3                                  .....
<= Recv SSL data, 1 bytes (0x1)
0000: 17                                              .
<= Recv data, 2994 bytes (0xbb2)
0000: 80 e2 94 80 e2 94 80 e2 94 80 e2 94 80 e2 94 80 ................
0010: e2 94 80 e2 94 80 e2 94 80 e2 94 80 e2 94 80 e2 ................
0020: 94 80 e2 94 80 e2 94 80 e2 94 80 e2 94 80 e2 94 ................
0030: 80 e2 94 80 e2 94 80 e2 94 80 e2 94 80 e2 94 80 ................
0040: e2 94 98 0a 20 20 20 20 20 20 20 20 20 20 20 20 ....            
0050: 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20                 
```

The UTF-8 byte sequence for a horizontal box drawing character is e2 94 80
which you can see a lot of here. But what's most relevant here is that the e2
94 comes in at the end of a frame, which is interrupted by some TLS bytes, then
it continues in another data from with the final 80 byte. Clearly, this
interruption is causing problems somewhere. But where? Is it Windows that
insists that all bytes in a UTF-8 triplet arrive in the same write call and
wine is faithfully emulating the corruption that occurs when it doesn't happen?
Since curl is writing the correct bytes to a file, it's probably not a curl problem per se.

**Comment by dfandrich** [comment]
Jay's Nov. 3 comment seems to confirm that this is basically what's happening and it's the same issue.

**Comment by OSPanel** [comment]
> Jay's Nov. 3 comment seems to confirm that this is basically what's happening and it's the same issue.

Maybe curl is better to just stop outputting data to the console until the entire response is received, because this leads to distortion when gluing bytes...

**Comment by dfandrich** [comment]
The output could be infinitely long, so it's not possible to wait for it all in
the general case.  Think about a longer or looping version of this, for
example: curl -s http://artscene.textfiles.com/vt100/movglobe.vt

**Comment by OSPanel** [comment]
Can anyone help with fixing this error, please. Jay hasn't had time for this in 5 months.

**Comment by jay** [comment]
I have some changes for this but I'm still working on it.
https://github.com/curl/curl/compare/master...jay:curl:utf8_win_console_fix?expand=1

**Comment by NativeDevMan** [comment]
Please tell me, if you have information, how soon will https://github.com/curl/curl/pull/10890 be accepted into the main branch?
